src/compiler/rust.rs

```diff
@@ -1184,6 +1427,239 @@ impl pkg::ToolchainPackager for RustToolchainPackager {
+                    return Ok(deps_detail.deps.clone())
+                }
+            }
+        }
+
+        trace!("Discovering dependencies of {}", rlib.display());
+
+        let mut cmd = process::Command::new(&self.executable);
+        cmd.args(&["-Z", "ls"]).arg(&rlib)
+            .env_clear()
+            .envs(ref_env(env_vars))
+            .env("RUSTC_BOOTSTRAP", "1"); // TODO: this is fairly naughty
+
+        let process::Output { status, stdout, stderr } = cmd.output()?;
+
+        if !status.success() {
+            bail!(format!("Failed to list deps of {}", rlib.display()))
+        }
+        if !stderr.is_empty() {
+            bail!("rustc -Z ls stderr non-empty: {:?}", String::from_utf8_lossy(&stderr))
+        }
+
+        let stdout = String::from_utf8(stdout).chain_err(|| "Error parsing rustc -Z ls output")?;
+        let deps: Vec<_> = parse_rustc_z_ls(&stdout)
+            .map(|deps| deps.into_iter().map(|dep| dep.to_owned()).collect())?;
+
+        {
+            // This will behave poorly if the rlib is changing under our feet, but in that case rustc
+            // will also do the wrong thing, so the user has bigger issues to deal with.
+            let mut cache = self.cache.lock().unwrap();
+            cache.insert(rlib.to_owned(), RlibDepsDetail { deps: deps.clone(), mtime: rlib_mtime });
+        }
+        Ok(deps)
+    }
+}
+
+// Parse output like the following:
+//
+// ```
+// =External Dependencies=
+// 1 std-08a5bd1ca58a28ee
+// 2 core-ed31c38c1a60e6f9
+// 3 compiler_builtins-6bd92a903b271497
+// 4 alloc-5184f4fa2c87f835
+// 5 alloc_system-7a70df28ae5ce6c3
+// 6 libc-fb97b8e8c331f065
+// 7 unwind-3fec89e45492b583
+// 8 alloc_jemalloc-3e9fce05c4bf31e5
+// 9 panic_unwind-376f1801255ba526
+// 10 bitflags-f482823cbc05f4d7
+// 11 cfg_if-cf72e166fff77ced
+// ```
+#[cfg(feature = "dist-client")]
+fn parse_rustc_z_ls(stdout: &str) -> Result<Vec<&str>> {
+    let mut lines = stdout.lines();
+    match lines.next() {
+        Some("=External Dependencies=") => {},
+        Some(s) => bail!("Unknown first line from rustc -Z ls: {}", s),
+        None => bail!("No output from rustc -Z ls"),
+    }
+
+    let mut dep_names = vec![];
+
+    while let Some(line) = lines.next() {
+        if line == "" {
+            break
+        }
+
+        let mut line_splits = line.splitn(2, ' ');
+        let num: usize = line_splits.next().expect("Zero strings from line split").parse()
+            .chain_err(|| "Could not parse number from rustc -Z ls")?;
+        let libstring = line_splits.next().ok_or_else(|| "No lib string on line from rustc -Z ls")?;
+        if num != dep_names.len() + 1 {
+            bail!("Unexpected numbering of {} in rustc -Z ls output", libstring)
+        }
+        assert!(line_splits.next().is_none());
+
+        let mut libstring_splits = libstring.rsplitn(2, '-');
+        // Rustc prints strict hash value (rather than extra filename as it likely should be)
```

Comment:
You filed this upstream, right? If so it might be good to link the issue in this comment.